salt/modules/x509.py

```diff
@@ -1464,12 +1476,41 @@ def create_certificate(
                 serial_number -= int(serial_number / sys.maxsize) * sys.maxsize
     cert.set_serial_number(serial_number)
 
+    # Handle not_before and not_after dates for custom certificate validity
+    fmt = '%Y-%m-%d %H:%M:%S'
+    if 'not_before' in kwargs:
+        time = datetime.datetime.strptime(kwargs['not_before'], fmt)
```

Comment:
Even though you specify in the docs what the format should be, I would prefer seeing this wrapped in a `try/except` to catch the failure to convert, and provide a human-friendly error message, e.g.

`'not_before: {!r} is not in required format %Y-%m-%d %H:%M:%S'.format(kwargs['not_before'])`

Otherwise, in the inevitable future when someone makes a typo in their `not_before`/`not_after`, they'll see an ugly stack trace, instead of a helpful error message that tells them how they goofed up.

I'm happy to be persuaded that this shouldn't be the case if you feel strongly about it, or if this is already happening somewhere else that I don't see in here. Otherwise, could you change that both here and in `not_after`?

Reply:
Having been bit countless times in the past by some very unfriendly stack traces on Salt I agree 100% with you. I think I got so used to this behaviour that it didn't even occur to me to have something friendlier there. I will gladly add the try/except to both `not_before` and `not_after`.

Reply:
@waynew I've added the exception handling code together with some unit tests to verify that it was really working. I'd appreciate if you could have a look at the unit tests `test_create_certificate_with_not_before_wrong_date` and `test_create_certificate_with_not_after_wrong_date` as well, specially on the message verification part for I see they might need an improvement as I'm not sure that the wrong string format will actually show up with a 'u' in it (e.g. u'2002/11/22').

Reply:
@waynew, hmmm I'm not sure of that. The problem wasn't the entire message showing up between `u''` but only the date passed in a wrong format. Also I did not change anything in the code but a sync to the latest code in branch was making the test fail as the date was no longer showing up between `u''`, as below:

```
======================================================================
FAIL: test_create_certificate_with_not_after_wrong_date (unit.modules.test_x509.X509TestCase)
[CPU:44.4%|MEM:30.2%] 
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jeduardo/src/salt-dev/tests/unit/modules/test_x509.py", line 457, in test_create_certificate_with_not_after_wrong_date
    self.assertEqual(msg, e.message)
AssertionError: u"not_after: u'this is an intentionally wrong format' is not in required format  [truncated]... != u'not_after: this is an intentionally wrong format is not in required format %Y- [truncated]...
- not_after: u'this is an intentionally wrong format' is not in required format %Y-%m-%d %H:%M:%S
?            --                                     -
+ not_after: this is an intentionally wrong format is not in required format %Y-%m-%d %H:%M:%S


======================================================================
FAIL: test_create_certificate_with_not_before_wrong_date (unit.modules.test_x509.X509TestCase)
[CPU:0.0%|MEM:30.2%] 
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jeduardo/src/salt-dev/tests/unit/modules/test_x509.py", line 396, in test_create_certificate_with_not_before_wrong_date
    self.assertEqual(msg, e.message)
AssertionError: u"not_before: u'this is an intentionally wrong format' is not in required format [truncated]... != u'not_before: this is an intentionally wrong format is not in required format %Y [truncated]...
- not_before: u'this is an intentionally wrong format' is not in required format %Y-%m-%d %H:%M:%S
?             --                                     -
+ not_before: this is an intentionally wrong format is not in required format %Y-%m-%d %H:%M:%S
```

So I assume something got fixed elsewhere in the meanwhile which is now causing this string to be properly ingested. 

Anyway, I've fixed the tests to check for the correct format of the message now.